modin/numpy/arr.py

```diff
@@ -704,50 +741,99 @@ def mean(self, axis=None, dtype=None, out=None, keepdims=None, *, where=True):
             if dtype is not None
             else (out.dtype if out is not None else self.dtype)
         )
+        out_type = out_dtype if not hasattr(out_dtype, "type") else out_dtype.type
+        if isinstance(where, array) and issubclass(out_type, numpy.integer):
+            out_dtype = numpy.float64
         check_kwargs(keepdims=keepdims, where=where)
         if self._ndim == 1:
             if axis == 1:
                 raise numpy.AxisError(1, 1)
-            result = self._query_compiler.astype(
-                {col_name: out_dtype for col_name in self._query_compiler.columns}
+            result = self._query_compiler
+            if isinstance(where, array):
+                result = where.where(self, numpy.nan)._query_compiler
+            result = result.astype(
+                {col_name: out_dtype for col_name in result.columns}
             ).mean(axis=0)
             if keepdims:
                 if out is not None and out.shape != (1,):
                     raise ValueError(
                         f"operand was set up as a reduction along axis 0, but the length of the axis is {out.shape[0]} (it has to be 1)"
                     )
-                return fix_dtypes_and_determine_return(result, 1, dtype, out, where)
+                if out is not None:
+                    out._query_compiler = (
+                        numpy.ones_like(out) * numpy.nan
+                    )._query_compiler
+                return (
+                    fix_dtypes_and_determine_return(
+                        result, 1, dtype, out, where is not False
+                    )
+                    if where is not False or out is not None
+                    else array([numpy.nan], dtype=out_dtype)
+                )
             return result.to_numpy()[0, 0] if where else numpy.nan
         if axis is None:
+            result = self
+            if isinstance(where, array):
+                result = where.where(self, numpy.nan)
             result = (
-                self.flatten()
-                .astype(out_dtype)
-                .mean(axis=axis, dtype=dtype, out=out, keepdims=None, where=where)
-            )
+                numpy.array(
+                    [result._query_compiler.sum(axis=1).sum(axis=0).to_numpy()[0, 0]],
+                    dtype=out_dtype,
+                )
+                / (
+                    prod(self.shape)
+                    - result._query_compiler.isna()
+                    .sum(axis=1)
+                    .sum(axis=0)
+                    .to_numpy()[0, 0]
+                )
```

Comment:
I wonder if extracting the calculation of "now many NAs is there" to a separate variable would improve the readability... my first impression was that this piece was accidentally calling `.sum()` on only part of the equation, until I realized what was going on

Reply:
I can break this up and add some comments to help explain what's going on!